Modernize FakeTensorMode, throw on non-fake inputs

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #78522
* #78677
* #78536
* #78090
* __->__ #78516

Modernizes FakeTensorMode by inheriting from `TorchDispatchMode`. `FakeTensor` and `FakeTensorMode` now call into a common helper function. I didn't see any existing idiomatic pattern for this so please feel free to comment if you think something else would be more ergonomic. 

This also throws if a non-Fake tensor is an input to `FakeTensorMode`.. So it effectively only extends handling to constructors (for now, more general handling later in stack). This is because we need more careful logic to detect something like:
```
x = torch.rand([1, 1])
with FakeTensorMode()
    y = x.add_(3)
    x.resize_([4])  # y should be resized here as well, no way to support this, error
```

Commit:
Modernize FakeTensorMode, throw on non-fake inputs

[ghstack-poisoned]

Commit:
Update on "Modernize FakeTensorMode, throw on non-fake inputs"




[ghstack-poisoned]

Commit:
Update on "Modernize FakeTensorMode, throw on non-fake inputs"




[ghstack-poisoned]

Commit:
Update on "Modernize FakeTensorMode, throw on non-fake inputs"




[ghstack-poisoned]

Commit:
Update on "Modernize FakeTensorMode, throw on non-fake inputs"


Modernizes FakeTensorMode by inheriting from `TorchDispatchMode`. `FakeTensor` and `FakeTensorMode` now call into a common helper function. I didn't see any existing idiomatic pattern for this so please feel free to comment if you think something else would be more ergonomic. 

This also throws if a non-Fake tensor is an input to `FakeTensorMode`.. So it effectively only extends handling to constructors (for now, more general handling later in stack).

[ghstack-poisoned]

ezyang commented on test/test_fake_tensor.py:
```diff
@@ -52,22 +52,26 @@ def test_type_as(self):
 
     def test_constructor(self):
-        with enable_torch_dispatch_mode(FakeTensorMode):
+        with enable_torch_dispatch_mode(FakeTensorMode(inner=None)):
```
can we just write all of these tests using `FakeTensorMode.push()`? Or maybe @samdow's patch has landed so `with FakeTensorMode()` works now

samdow replied:
Patch not landed yet (Richard has been hunting down a functorch/dispatch key bug and then I was going to ask him if he had anything to add)--and this actually doesn't work yet because `push` uses `push_torch_dispatch_mode` which causes the error Elias saw late last week (`Creating a new Tensor subclass FakeTensor but the raw Tensor object is already associated to a python object of type Tensor`)

On that note, I forgot to mention Richard debugged that and what's happening is that because push makes every mode have an inner mode (setting BaseTorchDispatchMode if there isn't one set), `detach` has a pyobj associated with it and we end up with this error). The patch fixes this because it removes BaseTorchDispatchMode but it will come up if we nest fake tensor mode with another mode. Long technical way of saying **if we want FakeTensorMode to be composable, we should use push and add `no_dispatch` around all constructors in `torch_dispatch`. I'll try to flag everywhere this will come up**

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -77,75 +162,9 @@ def __repr__(self):
-        return tree_map(partial(wrap, device=common_device), r)
+        def run_fn(func, types, args, kwargs):
+            return torch.Tensor.__torch_dispatch__(func, types, args, kwargs)
+        return torch_dispatch_impl(cls, func, types, args, kwargs, run_fn)
```
@samdow and I discussed this on Thursday and we think the right way to do code reuse is you put the real implementation in the mode, and then in the subclass implementation you (1) store the mode that allocated the subclass and (2) enable that mode before redispatching on the tensors as is.

eellison replied:
I'm going to leave to follow up because changing this causes a few different errors (which i'll file issues / smaller repros for and/or debug) and i would like to unblock dynamo

samdow commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -51,6 +52,90 @@ def _is_tensor_constructor(func: OpOverload):
+    # are not FakeTensors. For now, throw if any non-Fake Tensor inputs
+    # and just support constructors. TODO: extend more broadly
+    if isinstance(cls_or_mode_instance, FakeTensorMode):
+        conversion_made = False
+
+        def check_non_fake_tensor(x):
+            nonlocal conversion_made
+            conversion_made = conversion_made or (isinstance(x, torch.Tensor) and not isinstance(x, FakeTensor))
+
+        tree_map(check_non_fake_tensor, args)
+        tree_map(check_non_fake_tensor, kwargs)
```
nit: not sure which is more idiomatic but we could do `any(tree_flatten(tree_map...)[0])` or even something like
```
conversion_made = False
for x in tree_flatten(args)[0] + tree_flatten(kwargs)[0]:
  if isinstance(x, torch.Tensor) and not isinstance(x, FakeTensor):
    raise ...
```

this just saves us from using a nonlocal and lets us short circuit in the case that we run into the error

eellison replied:
IMO it's not worth optimizing for failure modes either in readability or runtime

eellison replied:
I'm mentally filing an issue to do `tree_iterate` separately or something along those lines because there are a lot of places in the codebase where `tree_map` is used where the return value isnt used

samdow commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -77,75 +162,9 @@ def __repr__(self):
-        common_device = FakeTensor._find_common_device(func, args, kwargs)
-
-        return tree_map(partial(wrap, device=common_device), r)
+        def run_fn(func, types, args, kwargs):
+            return torch.Tensor.__torch_dispatch__(func, types, args, kwargs)
```
We should be okay running `func(*args, **kwargs)` in all cases. This way we can also remove the `run_function` argument

eellison replied:
This actually causes an error: 
```
x = FakeTensor.from_tensor(torch.tensor(0.0))
y = FakeTensor.from_tensor(torch.rand([4, 4], device="cuda"))
out = x + y
...
ERROR: test_zero_dim (__main__.FakeTensorTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/test_fake_tensor.py", line 32, in test_zero_dim
    out = x + y
TypeError: unsupported operand type(s) for +: 'FakeTensor' and 'FakeTensor'
```

Although it's a mute point because I'm going to do the refactoring of tensors storing mode and calling into it

samdow commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -51,6 +52,90 @@ def _is_tensor_constructor(func: OpOverload):
+        out_device = new_kwargs.pop("device", torch.device("cpu"))
+        new_kwargs["device"] = torch.device("meta")
+        r = run_function(func, types, (), new_kwargs)
+        return FakeTensor(r, out_device)
```
should be in `no_dispatch` to allow for nested modes (more info https://github.com/pytorch/pytorch/pull/78516/files#r885905994)

samdow commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -51,6 +52,90 @@ def _is_tensor_constructor(func: OpOverload):
+    def wrap(e, device=None):
+        if isinstance(e, torch.Tensor) and not isinstance(e, FakeTensor):
+            if device:
+                return FakeTensor(e, device)
```
should be in `no_dispatch` to allow for nested modes (more info https://github.com/pytorch/pytorch/pull/78516/files#r885905994)

samdow reviewed (approved):
LGTM! All of my points are small/cleanup stuff

eellison committed:
Update on "Modernize FakeTensorMode, throw on non-fake inputs"


Modernizes FakeTensorMode by inheriting from `TorchDispatchMode`. `FakeTensor` and `FakeTensorMode` now call into a common helper function. I didn't see any existing idiomatic pattern for this so please feel free to comment if you think something else would be more ergonomic. 

This also throws if a non-Fake tensor is an input to `FakeTensorMode`.. So it effectively only extends handling to constructors (for now, more general handling later in stack). This is because we need more careful logic to detect something like:
```
x = torch.rand([1, 1])
with FakeTensorMode()
    y = x.add_(3)
    x.resize_([4])  # y should be resized here as well, no way to support this, error
```

[ghstack-poisoned]
